### ISSUE TITLE
Suggestion: Safe implicit default assignment operator ?=

### ISSUE THREAD

**Issue by frevds** [Microsoft/TypeScript#27860]
I'm running into this case very often where I have to use a dictionary resp. object as a registry, having to add keys if they don't exist and increment or modify the value behind, where I then have to distinguish two cases - either the key was in the dictionary (then I have to increment the value) or the key was not yet in it (then I have to assign the default value). 
This case also occurs for scalar values where one has to assign a default. 

My projects are littered with code like this: 
```
const foo = {}; 
if (!foo.hasOwnProperty('key'))
    foo['key'] = 0; 
foo['key']++; 
```
or this (arrays as values): 
```
const bah = {}; 
if (!bah.hasOwnProperty('key'))
    bah['key'] = []; 
bah['key'].push('foo'); 
```

Now it would be great to have some kind of shorthand for this pattern. 

It is somewhat **like a safe navigation operator that implicitly assigns a default** to the left-hand side, if and only if the left-hand side is not yet defined. 

It should be **chainable**, which highly improves code readability in case you have several levels of key-value registries. 

So instead of writing this: 
```
const registry = {}; 
if (!registry.hasOwnProperty('keyLevel1'))
    registry['keyLevel1'] = {}; 
if (!registry['keyLevel1'].hasOwnProperty('keyLevel2'))
    registry['keyLevel1']['keyLevel2'] = {}; 
registry['keyLevel1']['keyLevel2'].foo = 'bar'; 
```
I would want to just write and see something like this instead: 
```
((registry['keyLevel1'] ?= {})['keyLevel2'] ?= {}).foo = 'wow'; 
```
Here the ?= operator checks for whether the left-hand side value is null or undefined (so it would probe the indexing of the registry object in this case) and if so, it would assign the right-hand side (in this case it would assign the member keyLevel1 of the registry object the value {}). 

This should also work for scalars and values, not just for properties of objects (resp. keys in dictionaries): 
```
let a: IFoo = null; 
( a ?= new IFoo() ).wow(); 

let n: number = null; 
(n ?= 1) *= -2;  // brackets only for indication, should have precedence

const c: IFoo = { foo: null }; 
c.foo ?= 123;  // assign 123 but only if it does not have a value yet (excluding 0/NaN). 
```
For the above I would currently have to write: 
```
( a || (a = new IFoo()) ).moo();  // this check works for objects and arrays only

if ((n === null) || (n === undefined))  // checks for scalars and strings
    n = 1; 
n *= 2; 

if ((c.foo === null) || (c.foo === undefined))
    c.foo = 123; 
```
The Boolean coercion for scalars and strings will return false not only when null or undefined, but also when asking `if (0)` or `if (NaN)` or `if ('')`, making things error-prone and cumbersome to write. 

Not sure whether a mechanism for this exists already, or whether it exists in any language (in C# we have at least a safe navigation for members using the ?. operator as well as the ?? operator for a null check, but no implicit assignment operator afaik). 

Would be very handy to have though..

**Comment by mattmccutchen** [comment]
Did you read the following in the suggestion template?

> * [ ] This isn't a runtime feature (e.g. new expression-level syntax)

Essentially all runtime features are considered out of scope of TypeScript nowadays for fear that a conflicting implementation might be added to the ECMAScript standard in the future.

**Comment by frevds** [comment]
Oh, that must have escaped my eyes (these templates are hard to read in the editor view). 

> Essentially all runtime features are considered out of scope 
> for fear that a conflicting implementation might be added 
> to the ECMAScript standard in the future.

I see. Right, there shouldn't be conflicting syntax in TS and ES. Too bad for new features though, how can you ever improve TS over ES if it's not possible to introduce anything but keywords? Features are usually added to ES only after appearing and being used in languages such as TS. 

An keyword-based approach would look like this:

**Comment by frevds** [comment]
FYI Here is a solution without any language-level additions, for whoever needs it: 

Sticking to the original demand of initializing keys in dictionaries 
(resp. members in an object or entries in an array) there is an approach: 

```
const registry = {}; 
registry.ensure('keyLevel1', {}).ensure('keyLevel2', {}).foo = 'wohoo'; 
```

This requires the following prototype extension of any objects by a generic method, straight-forward: 
```
Object.prototype.ensure = function(memberKey, defaultValue) { if (!this.hasOwnProperty(memberKey)) this[memberKey] = defaultValue; return this[memberKey]; }; 
```

Or for TypeScript and optimized: 
```
declare interface Object {
    /** Safe member access (adds the specified key if not present and assigns the specified default value). Returns the member's value. */
    ensure<TValue>(memberKey: any, defaultValue: TValue): TValue; 
}

Object.prototype.ensure = function <TValue>(memberKey: any, defaultValue?: TValue): TValue { 
    return this.hasOwnProperty(memberKey) ? this[memberKey] : (this[memberKey] = defaultValue); 
}; 
```
This goes into an Ensure.ts to be imported like this: 
```
import './Ensure'
```


That suits my needs perfectly. It does not extend beyond objects to assign defaults to scalars but who cares. 
So no feature required (even though it would look better to have an operator for that).
